WIP: support multiple gitlab instances

tmgordeeva opened the pull request:
Refactors the script to support multiple gitlab instances.  As before, the
prefix argument points to our runner's eventual config.toml.  Now the templates
argument points to a directory of toml files, one per runner.  Runner toml
files are processed as before, but admin and access tokens need to be provided
per runner as each runner could use a different gitlab url.  Admin/access
tokens are not however persisted in the final config.toml generated.

Commit:
WIP: support multiple gitlab instances

Refactors the script to support multiple gitlab instances.  As before, the
prefix argument points to our runner's eventual config.toml.  Now the templates
argument points to a directory of toml files, one per runner.  Runner toml
files are processed as before, but admin and access tokens need to be provided
per runner as each runner could use a different gitlab url.  Admin/access
tokens are not however persisted in the final config.toml generated.

tmgordeeva commented:
@tgmachina the current draft I'm working on - this has some issues I'm fixing up to get it actually working, but a quick overview would be good in case there's anything major missing.

tgmachina commented on gitlab_runner_config.py:
```diff
@@ -200,113 +263,35 @@ def update_runner_config(config_template, config_file, internal_config):
-        """,
+        "-t",
+        "--templates",
+        default="/etc/gitlab-runner/main",
```
Rather than specify this as different path altogether, let's assume that whatever template directory is being used here exists under `prefix`. I would also suggest renaming to `instance` or something like that.

This will allow for a simpler invocation of `systemd`. The execution line would change to:

`ExecStartPre=gitlab_runner_auth.py --prefix /etc/gitlab-runner --instance %i`

And the instance would come from invoking `systemctl` to start another process:

`systemctl start gitlab-runner@bar.service # bar would populate %i above`.

In the absence of invoking with `@`, the default behavior would be to look for a folder `main` so the default arg route you have above would work for that (sans absolute path).

See this post for more details: https://stackoverflow.com/a/48232848

tgmachina commented on gitlab_runner_config.py:
```diff
@@ -200,113 +263,35 @@ def update_runner_config(config_template, config_file, internal_config):
+    runners = {}
 
-    update_runner_config(config_template, config_file, runner_config)
+    for filename in os.listdir(templates):
```
As an added measure of safety, it'd be good to ensure that the template directory only has owner permissions. I wrote that up in my initial refactor work--feel free to take it: https://github.com/LLNL/gitlab-runner-auth/blob/feature/multi-instance/gitlab_runner_config.py#L212

tgmachina commented on gitlab_runner_config.py:
```diff
@@ -62,111 +190,46 @@ def which(cmd):
+            else:
+                token = data.get("token", "")
+                # TODO: changed no longer needed?
+                config, _ = client.update_runner_token(token, runner_type)
```
I think your TODO is right. The token should come back in the request to list runners with this host's tags so it can just get pulled from `data` as you're doing.

tgmachina reviewed (commented):
All-in-all, I think you're on the right track and made very reasonable assumptions.

I've written in some comments and my other suggestion would be to break `read_runners` into smaller, more focused methods. I saw you were doing some of that already with the API so I imagine it's already something you're thinking about. I also recognize it's my fault that it was written that way in the first place 😅

Commit:
WIP: permissions checking, cleanup

Commit:
Gitlab client refactor

Splits out HTTP requests and errors.

Commit:
Split out requester in client

Commit:
Client unit tests

Additional unit tests for the gitlab client, some small fixes found through testing.

tgmachina commented:
Thank you for your effort on this @tmgordeeva. I took inspiration from some of the changes you made and leveraged [zipapp](https://docs.python.org/3/library/zipapp.html) for packaging, removing the need for a custom client. That said, I'll be closing this PR.
